src/background/vmModules/ApprovalController.ts

```diff
@@ -0,0 +1,253 @@
+      }) => {
+        if (updatedAction.id !== actionId) {
+          return;
+        }
+
+        if (error || newStatus === ActionStatus.ERROR) {
+          cleanup();
+          resolve({ error: rpcErrors.internal(error) });
+          return;
+        }
+
+        if (newStatus === ActionStatus.ERROR_USER_CANCELED) {
+          cleanup();
+          resolve({ error: providerErrors.userRejectedRequest() });
+          return;
+        }
+
+        if (newStatus === ActionStatus.SUBMITTING) {
+          try {
+            const { signedTx: signedData, txHash } = await this.#handleApproval(
+              params,
+              updatedAction,
+              network
+            );
+
+            if (signedData) {
+              resolve({ signedData });
+            } else if (txHash) {
+              resolve({ txHash });
+            } else {
+              resolve({
+                error: rpcErrors.internal({
+                  message: 'Unsupported signing result type',
+                }),
+              });
+            }
+          } catch (err) {
+            resolve({
+              error: rpcErrors.internal({
+                message: 'Unable to sign the message',
+                data: {
+                  originalError: err,
+                },
+              }),
+            });
+          } finally {
+            cleanup();
+          }
+
+          return;
+        }
+      };
+
+      actionsService.addListener(ActionsEvent.MODULE_ACTION_UPDATED, onUpdated);
+    });
+  };
+
+  #getPopupUrl = (method: RpcMethod) => {
+    if (method === RpcMethod.BITCOIN_SEND_TRANSACTION) {
+      return 'approve/bitcoinSignTx';
+    }
+
+    throw rpcErrors.methodNotSupported({
+      data: method,
+    });
+  };
+
+  #try<F extends (...args: any) => any>(
+    fn: F
+  ): [ReturnType<F>, null] | [null, RpcError] {
+    try {
+      return [fn(), null];
+    } catch (err: any) {
+      const safeError =
+        err instanceof JsonRpcError
+          ? err
+          : rpcErrors.internal({ message: 'Unknown error', data: err });
+      return [null, safeError];
+    }
+  }
+
+  #handleApproval = async (
+    params: ApprovalParams,
+    action: Action,
+    network: NetworkWithCaipId
+  ) => {
+    const { signingData } = action;
+
+    if (signingData?.type === RpcMethod.BITCOIN_SEND_TRANSACTION) {
+      this.#requestsMetadata.set(params.request.requestId, {
+        txType: 'send',
+        chainId: network.chainId,
+        rpcUrl: network.rpcUrl,
+        site: new URL(params.request.dappInfo.url).hostname,
+      });
+
+      measureDuration(params.request.requestId).start();
+
+      const { inputs, outputs } = await buildBtcTx(
+        signingData.account,
+        getProviderForNetwork(network) as BitcoinProvider,
+        {
+          amount: signingData.data.amount,
+          address: signingData.data.to,
+          feeRate: signingData.data.feeRate,
+          token: signingData.data.balance,
+        }
+      );
```

Comment:
The module already does it, but we do it here again in case user updates the fee rate on the approval screen. Ideally, we need a way to communicate changes like that back to the VM module and get an updated unsigned tx in return.

Here's a follow up ticket:

* https://ava-labs.atlassian.net/browse/CP-9159